dbos/_sys_db.py

```diff
@@ -1596,17 +1601,39 @@ def get_event(
             )
         return value
 
-    def enqueue(self, workflow_id: str, queue_name: str, conn: sa.Connection) -> None:
+    def enqueue(
+        self,
+        workflow_id: str,
+        queue_name: str,
+        conn: sa.Connection,
+        *,
+        enqueue_options: Optional[EnqueueOptionsInternal],
+    ) -> None:
         if self._debug_mode:
             raise Exception("called enqueue in debug mode")
-        conn.execute(
-            pg.insert(SystemSchema.workflow_queue)
-            .values(
-                workflow_uuid=workflow_id,
-                queue_name=queue_name,
+        try:
+            deduplication_id = (
+                enqueue_options["deduplication_id"]
+                if enqueue_options is not None
+                else None
             )
-            .on_conflict_do_nothing()
-        )
+            conn.execute(
+                pg.insert(SystemSchema.workflow_queue)
+                .values(
+                    workflow_uuid=workflow_id,
+                    queue_name=queue_name,
+                    deduplication_id=deduplication_id,
+                )
+                .on_conflict_do_nothing(
+                    index_elements=SystemSchema.workflow_queue.primary_key.columns
+                )  # Ignore primary key constraint violation
+            )
+        except DBAPIError as dbapi_error:
+            # Unique constraint violation for the deduplication ID
+            if dbapi_error.orig.sqlstate == "23505":  # type: ignore
+                raise DBOSQueueDeduplicatedError(
+                    workflow_id, queue_name, deduplication_id
+                )
```

Comment:
I don't think we should raise an exception here. From the client standpoint, I'll probably write something like:

```
func event_receiver(event):
    time_dedup_id = current minute
    with setEnqueueOption(time_dedup_ip)
        queue.enqueue(event)
```

With the goal of only enqueuing one of these events a minute

I don't want the code to throw exceptions. I merely want to observe how much dedup happened.

Reply:
After discussion, we think throwing exceptions makes it easier to handle deduplicated scenarios, for example retry on `DBOSQueueDeduplicatedError` after a waiting period.